go-controller/pkg/ovn/loadbalancer/loadbalancer.go

```diff
@@ -36,41 +36,58 @@ import (
 // It is assumed that names are meaningful and somewhat stable, to minimize churn. This
 // function doesn't work with Load_Balancers without a name.
 func EnsureLBs(nbClient libovsdbclient.Client, externalIDs map[string]string, LBs []LB) error {
-	lbCache, err := GetLBCache(nbClient)
+	existing := libovsdbops.FindLoadBalancersByExternalIDs(nbClient, externalIDs)
+	existingByName := make(map[string]nbdb.LoadBalancer, len(existing))
+	toDelete := sets.NewString()
+
+	// gather some data
+	lbToSwitchNames, switchNamesToUUID, err := listSwitches(nbClient)
 	if err != nil {
-		return fmt.Errorf("failed initialize LBcache: %w", err)
+		return err
+	}
+	lbToRouterNames, routerNamesToUUID, err := listRouters(nbClient)
+	if err != nil {
+		return err
+	}
+	lbToGroupNames, groupNamesToUUID, err := listGroups(nbClient)
+	if err != nil {
+		return err
 	}
-
-	existing := lbCache.Find(externalIDs)
-	existingByName := make(map[string]*CachedLB, len(existing))
-	toDelete := sets.NewString()
 
 	for _, lb := range existing {
+		// collision - somehow things didn't come together how we expected
+		// rare but we have seen it happen, see https://bugzilla.redhat.com/show_bug.cgi?id=2042001
+		if _, ok := existingByName[lb.Name]; ok {
+			klog.V(2).Infof("Name collision for load balancer %s: deleting both and re-creating")
+			toDelete.Insert(lb.UUID)
+			delete(existingByName, lb.Name)
+			continue
+		}
 		existingByName[lb.Name] = lb
 		toDelete.Insert(lb.UUID)
 	}
 
-	lbs := make([]*nbdb.LoadBalancer, 0, len(LBs))
+	desired := make([]*nbdb.LoadBalancer, 0, len(LBs))
+
 	addLBsToSwitch := map[string][]*nbdb.LoadBalancer{}
 	removeLBsFromSwitch := map[string][]*nbdb.LoadBalancer{}
 	addLBsToRouter := map[string][]*nbdb.LoadBalancer{}
 	removesLBsFromRouter := map[string][]*nbdb.LoadBalancer{}
 	addLBsToGroups := map[string][]*nbdb.LoadBalancer{}
 	removeLBsFromGroups := map[string][]*nbdb.LoadBalancer{}
-	wantedByName := make(map[string]*LB, len(LBs))
-	for i, lb := range LBs {
-		wantedByName[lb.Name] = &LBs[i]
+
+	for _, lb := range LBs {
 		blb := buildLB(&lb)
-		lbs = append(lbs, blb)
-		existingLB := existingByName[lb.Name]
+		desired = append(desired, blb)
+		existingLB, exists := existingByName[lb.Name]
 		existingRouters := sets.String{}
 		existingSwitches := sets.String{}
 		existingGroups := sets.String{}
-		if existingLB != nil {
+		if exists {
```

Comment:
Can we add the UUID to blb from existingLB if it exists? Or is it being done already somehow and I am missing it?

Reply:
Definitely. Good point.